x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchStoreService.java

```diff
@@ -0,0 +1,165 @@
+import java.util.Random;
+
+import static org.elasticsearch.xpack.core.ClientHelper.INDEX_LIFECYCLE_ORIGIN;
+import static org.elasticsearch.xpack.search.AsyncSearchTemplateRegistry.INDEX_TEMPLATE_VERSION;
+import static org.elasticsearch.xpack.search.AsyncSearchTemplateRegistry.ASYNC_SEARCH_TEMPLATE_NAME;
+import static org.elasticsearch.xpack.search.TransportGetAsyncSearchAction.ensureAuthenticatedUserIsSame;
+
+/**
+ * A class that encapsulates the logic to store and retrieve {@link AsyncSearchResponse} to/from the .async-search index.
+ */
+class AsyncSearchStoreService {
+    private static final Logger logger = LogManager.getLogger(AsyncSearchStoreService.class);
+
+    static final String ASYNC_SEARCH_ALIAS = ASYNC_SEARCH_TEMPLATE_NAME + "-" + INDEX_TEMPLATE_VERSION;
+    static final String RESULT_FIELD = "result";
+    static final String HEADERS_FIELD = "headers";
+
+    private final Client client;
+    private final NamedWriteableRegistry registry;
+    private final Random random;
+
+    AsyncSearchStoreService(Client client, NamedWriteableRegistry registry) {
+        this.client = new OriginSettingClient(client, INDEX_LIFECYCLE_ORIGIN);
+        this.registry = registry;
+        this.random = new Random(System.nanoTime());
+    }
+
+    /**
+     * Store an empty document in the .async-search index that is used
+     * as a place-holder for the future response.
+     */
+    void storeInitialResponse(Map<String, String> headers, ActionListener<IndexResponse> listener) {
+        IndexRequest request = new IndexRequest(ASYNC_SEARCH_ALIAS)
+            .id(UUIDs.randomBase64UUID(random))
+            .source(Collections.singletonMap(HEADERS_FIELD, headers), XContentType.JSON);
+        client.index(request, listener);
+    }
+
+    /**
+     * Store the final response if the place-holder document is still present (update).
+     */
+    void storeFinalResponse(Map<String, String> headers, AsyncSearchResponse response,
+                            ActionListener<UpdateResponse> listener) throws IOException {
+        AsyncSearchId searchId = AsyncSearchId.decode(response.id());
+        Map<String, Object> source = new HashMap<>();
+        source.put(RESULT_FIELD, encodeResponse(response));
+        source.put(HEADERS_FIELD, headers);
+        UpdateRequest request = new UpdateRequest().index(searchId.getIndexName()).id(searchId.getDocId())
+            .doc(source, XContentType.JSON)
+            .detectNoop(false);
+        client.update(request, listener);
+    }
+
+    /**
+     * Get the response from the .async-search index if present, or delegate a {@link ResourceNotFoundException}
+     * failure to the provided listener if not.
+     */
+    void getResponse(AsyncSearchId searchId, ActionListener<AsyncSearchResponse> listener) {
+        final Authentication current = Authentication.getAuthentication(client.threadPool().getThreadContext());
+        GetRequest internalGet = new GetRequest(searchId.getIndexName())
```

Comment:
Because the result index is encoded in the asyncTaskId, a user can create his own index, and then craft an asyncTaskId that the target user can use to get a bogus result.

Reply:
Is is an issue ? The index name is resolved in the submit action so changing the index when retrieving an id should not impact the storage of the "real" response. In what scenario do you see this happening ? The worst case is that we return the "fake" response of the other index but we'll never return the "real" response. We can also add a validation that the index name matches the async-search alias since it's a protected namespace. Would that solve your concern ?

Reply:
It is not a big problem in the same sense that users can alter one another 's responses, but being able to craft responses is still problematic.

I prefer we offer users the following guarantee: If you call the get response API you get the response to a job that you submitted, OR an error response (missing) if the job is expired or otherwise missing (node restarted) or if the job-id is not yours. Somebody else should not be able to craft responses that you interpret as valid.

Does it make sense?

Reply:
Sure, that makes sense. I can add a validation that checks if the index name matches the async search index pattern. That would be enough imo, do you agree ?

Reply:
Yes, I think that would work.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/baa498d84d71d87c4afad294845c3c4f68853672 to validate the index name and to fail the request if the index name does not match the index pattern.

Reply:
I pushed the [validation](https://github.com/elastic/elasticsearch/pull/49931/commits/d03b9402192b03a592e08907e07b9ac10f5f5515) in the decoding of the search id, this means that it will be checked on all APIs.